### PR TITLE
Retrieve all known credentials for DockerBuildImage custom task

### LINKED ISSUE #908: Retrieve all known credentials for DockerBuildImage custom task
The `DockerBuildImage` task uses registry credentials that are configured on the task/project configuration.

Recently, `DockerPushImage` and `DockerPullImage` gained the ability to read credentials from the native `docker login` authentication, but `DockerBuildImage` does not yet support this and only uses credentials configured inside Gradle.

I am willing to contribute this change, but I first wanted to open an issue to have some feedback from you before I start implementing it.

### Expected Behavior

Similar to `DockerPushImage` & `DockerPullImage`, `DockerBuildImage` also uses `docker login` credentials when it needs to authenticate to pull images.

For `DockerBuildImage`, multiple credentials can be sent at the same time, so the application does not need to know which credentials it needs. (there can be multiple `FROM <image>` in a `Dockerfile` going to different private registries)

The `RegistryAuthLocator` will gather a list of all credentials and will place them all inside `AuthConfigurations` in this block of code: https://github.com/bmuschko/gradle-docker-plugin/blob/7e725b65aea32d8fa77dd33e2a579941694c395c/src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy#L270-L273

How to discover all available credentials is something that needs to be researched a bit, but I think reading the `"auths"` section from `~/.docker/config.json`, combined with the hostnames discovered from `docker-credential-?? list` should give us a list of all the available credentials to send.

### Current Behavior

Currently, `DockerBuildImage` only uses the `registryCredentials` configured on the task to authenticate.

### Context

We are building private docker images on top of other private docker images (all in the same registry). To make sure we are always using the latest version of the images, we set `pull = true` on `DockerBuildImage`, which always pulls the `FROM ???` images configured in a Dockerfile.

Currently, we configure `registryCredentials` with credentials read from `~/.gradle/gradle.properties` on each developers computer. We want to get rid of this file with plain text credentials, and instead delegate this credential management to docker.

### PR DESCRIPTION
**Pull request by vierbergenlars**
Extend RegistryAuthLocator to retrieve all credentials known in .docker/config.json and the credentials store.

Fixes #908

### CODE REVIEW

**Comment by vierbergenlars** [comment]
I think the test failure is not related to my changes, as the failing test passes locally. (though an other one fails there, as described in #914 )

**Comment by vierbergenlars** [comment]
I manually ran the commands that should be equivalent with the failing test.

It fails because the account `postgres` is now disabled in the container, resulting in `su postgres [...]` failing with exit code 1, and not stopping the container.

https://github.com/bmuschko/gradle-docker-plugin/blob/0b925626ea69159382316a80344d5f6240a12f65/src/functTest/groovy/com/bmuschko/gradle/docker/tasks/container/DockerExecStopFunctionalTest.groovy#L15

It should work by replacing it with the equivalent of `docker exec -u postgres cc74287d88577e1b21b9b0a0beadb09a7b34bcaa844fbe0d5e33feedd21e66e3  /usr/local/bin/pg_ctl stop -m fast`

**Comment by bmuschko** [comment]
Would you mind rebasing your changes on `master`?

File: src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy [hunk 1]
```diff
@@ -130,16 +139,112 @@ class RegistryAuthLocator {
...
+    }
+
+    /**
+     * Gets all authorization information
+     * using $DOCKER_CONFIG/.docker/config.json file
+     * If missing, an empty AuthConfigurations object is returned
+     * @return AuthConfigurations object containing all authorization information,
+     * or an empty object if not available
+     */
+    AuthConfigurations lookupAllAuthConfigs() {
```
Comment (bmuschko) [inline src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy]: Shouldn't we be able to reuse logic from `lookupAuthConfig`? Seems like we are duplicating logic.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy [hunk 2]
```diff
@@ -267,9 +267,7 @@ class DockerBuildImage extends AbstractDockerRemoteApiTask implements RegistryCr
             buildImageCmd.withTarget(target.get())
         }
 
-        AuthConfigurations authConfigurations = new AuthConfigurations()
-        AuthConfig authConfig = getRegistryAuthLocator().createAuthConfig(registryCredentials)
-        authConfigurations.addConfig(authConfig)
+        AuthConfigurations authConfigurations = getRegistryAuthLocator().lookupAllAuthConfigs(registryCredentials);
```
[the remark this answers is not in the document]
Comment (vierbergenlars) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: To add a functional test for this, we would need to modify the user's `.docker/config.json` or their credentialsstore.
`DockerPullImage`/`DockerPushImage` also don't have functional tests for this part.

If you insist on a functional test, can we look to add a test that only runs on CI, as to not to mess with the environment of a developer that runs the tests?

File: src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy [hunk 3]
```diff
@@ -130,16 +139,112 @@ class RegistryAuthLocator {
...
-        defaultAuthConfig
+        return null
+    }
+
+    /**
+     * Gets all authorization information
+     * using $DOCKER_CONFIG/.docker/config.json file
+     * If missing, an empty AuthConfigurations object is returned
+     * @return AuthConfigurations object containing all authorization information,
+     * or an empty object if not available
```
[the remark this answers is not in the document]
Comment (vierbergenlars) [reply src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy]: Linebreaks like this are already present in the code (https://github.com/bmuschko/gradle-docker-plugin/pull/913/files#diff-60a636e371b6c5cd20fffa134ee90abdR93-R94)
I think this was done automatically by the InteliJ formatter.

File: src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/tasks/image/DockerBuildImage.groovy]: We do have some tests for pushing though there could be more: https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/functTest/groovy/com/bmuschko/gradle/docker/DockerJavaApplicationPluginFunctionalTest.groovy#L226-L272
These tests rely on a private registry set up on CI which you could set up locally as well if needed. I think those tests should rather be based on the custom tasks than the convention plugins. Given that you are the main user of this change, we'll probably hear back from you if it doesn't work. ;-)

File: src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy (hunk 1 above)
Comment (bmuschko) [reply src/main/groovy/com/bmuschko/gradle/docker/internal/RegistryAuthLocator.groovy]: I might look into additional refactorings later.

**Comment by bmuschko** [comment]
Thanks for contributing. This has been merged. Sorry it took so long.

**Comment by vierbergenlars** [comment]
No problem, thanks for your plugin and time!

**Pull request merged**